core/modules/filter/filter.admin.inc

```diff
@@ -484,7 +476,13 @@ function filter_admin_format_filter_settings_form_ajax($form, &$form_state) {
   else {
     $trigger = $form_state['triggering_element'];
     $name = $trigger['#name'];
-    $format = $form_state['format'];
+    // Load previously added settings from tempstore, if any.
+    if ($stored_format = filter_get_format_tempstore($form_state['format']->format)) {
+      $format = $stored_format;
+    }
+    else {
+      $format = $form_state['format'];
+    }
```

Comment:
This part doesn't seem quite right. The form `#submit` handlers will fire _before_ any `#ajax` handlers, so I think any changes to `$form_state['format']` should be in `filter_admin_format_filter_settings_form_redirect()`, and `$form_state['format']` should be up-to-date in this function.

Reply:
Hm... This only _gets_ the values, it doesn't set them. Update of tempstore happens in filter_admin_format_filter_settings_update, which is the submit handler.

Reply:
Sorry I have trouble outlining the logic flow. What should happen here in non-AJAX scenarios is:

- The overall format form is built.
- Form is submitted (`filter_admin_format_filter_settings_form_redirect()`)
  - Tempstore is set with changed values.
  - User is redirected.

In AJAX scenarios it's nearly the same, except an additional function (through `#ajax`) is added on, which opens the dialog instead of redirecting:
- The overall format form is built.
- Form is submitted (`filter_admin_format_filter_settings_form_redirect()`)
  - Tempstore is set with changed values
- Dialog is opened via AJAX command (`filter_admin_format_filter_settings_form_ajax()`).

In both cases `filter_admin_format_filter_settings_form_redirect()` is called, which is where I would expect the tempstore to be set.

The part I'm not understanding here is that you've remove `filter_set_format_tempstore()` from `filter_admin_format_filter_settings_form_redirect()` so tempstore isn't updated in non-AJAX scenarios before editing the filter settings. What I would have expected is that `filter_set_format_tempstore()` is set in the `filter_admin_format_filter_settings_form_redirect()` function always. Anything modified in `$form_state['format']` would get passed to `filter_admin_format_filter_settings_form_ajax()` within the same request, so you shouldn't need to load it again.

Reply:
Okay, I think I understand what's happening here a little now.

`$form_state['format']` is being stored not only in the tempstore as the format, but also in the Form API cache. This means that some changes to `$form_state['format']` happen automatically and we don't always have to load from the tempstore.